napari/layers/shapes/_shape_list.py

```diff
@@ -860,15 +879,25 @@ def to_colors(self, colors_shape=None, zoom_factor=1, offset=[0, 0]):
         colors = np.zeros(tuple(colors_shape) + (4,), dtype=float)
         colors[..., 3] = 1
 
-        for ind in self._z_order[::-1]:
-            if self._displayed[ind]:
-                mask = self.shapes[ind].to_mask(
-                    colors_shape, zoom_factor=zoom_factor, offset=offset
-                )
-                if type(self.shapes[ind]) in [Path, Line]:
-                    col = self._edge_color[ind]
-                else:
-                    col = self._face_color[ind]
-                colors[mask, :] = col
+        z_order = self._z_order[::-1]
+        shapes_in_view = np.argwhere(self._displayed)
+        z_order_in_view_mask = np.isin(z_order, shapes_in_view)
+        z_order_in_view = z_order[z_order_in_view_mask]
+
+        if max_shapes is not None and len(z_order_in_view) > max_shapes:
+            indices_to_use = np.random.randint(
+                0, len(z_order_in_view), max_shapes
+            )
+            z_order_in_view = z_order_in_view[indices_to_use]
```

Comment:
```suggestion
            z_order_in_view = z_order_in_view[:max_shapes]
```

I think we should actually take to top shapes here (syntax could be with `z_order_in_view[-max_shapes:]`) instead, you have to check.

Reason is that right now every time i drag a shape around the thumbnail picks a random 100, which looks really funny - see

![shapes_thumbnails_changing](https://user-images.githubusercontent.com/6531703/91106019-7c7e0200-e626-11ea-98a2-0307543462a2.gif)

Reply:
Hey @sofroniewn , that's a great point. I've made the suggested change.